[Shin Jin] iP

jinnieshin opened the pull request.

Commit:
Add Gradle support

jinnieshin committed:
Level 1

Implemented Greet, Echo, Exit

jinnieshin committed:
Level 2

Implemented Add, List

jinnieshin committed:
Level 3

Implemented Mark as Done

jinnieshin committed:
Level 4

Implemented ToDos, Deadlines, Events

jinnieshin committed:
Automated Text UI Testing

Test using the I/O redirection technique

jinnieshin committed:
Level 6, A-Collections

Implemented Delete, used java Collections

jinnieshin committed:
Level 5

Implemented Handle Errors

jinnieshin committed:
Level 7

Implemented saving and loading of tasks added into Duke.

Saving and loading of the tasks is required to remember the tasks added even after the user exits Duke.

Used Java's I/O system to read from the file that saves data, named 'duke.txt'.

Each line of the text file represents a task's type, status, title, followed by deadline or duration, where each components is separated by '|'. Duration has a format of (from)-(to).

jinnieshin committed:
Level 8

Used LocalDateTime for variables handling date and time.

Date and time of the deadline and event can be used more meaningfully.

jinnieshin committed:
clear git cache

jinnieshin committed:
Merge pull request #1 from jinnieshin/branch-Level-8

Level 8

jinnieshin committed:
Level 8 update

Fixed saving format of date and time of tasks

jinnieshin committed:
AMoreOOP

Changed structure of Duke in a more OOP style.

Divided main function in Duke into classes Ui, Storage, and Parser.

Created an abstract class called Command and created children of Command for each command input by the user.

This was done to each Unit Testing of the software.

jinnieshin committed:
AMoreOOP (2)

Finished implementing and checking of Ui, Parser and Commands for Duke to be functioning in the correct way.

Implemented FormatChecker to Check formats for different String inputs that needs to be formatted to be processed.

jinnieshin committed:
Merge pull request #3 from jinnieshin/branch-AMoreOOP

Branch a more oop

jinnieshin committed:
Update .gitignore

Hide data.txt to protect user information

jinnieshin committed:
clear git cache

jinnieshin committed:
A-Packages

Organised classes into suitable java packages.

Largest package is duke, commands are in duke.command, tasks(todo, deadline, event) are in duke.task.

jinnieshin committed:
Update .gitignore

jinnieshin committed:
clear git cache

jinnieshin committed:
Fixed bugs

Corrected date format, fixed minor bugs

jinnieshin committed:
Merge pull request #4 from jinnieshin/add-gradle-support

Add Gradle support

jinnieshin committed:
A-JUnit

Added JUnit Tests for FormatChecker and Command to test whether methods in FormatChecker and Command classes function correctly.

jinnieshin committed:
Updated data storage file path

jinnieshin committed:
A-Jar

Packaged duke as an executable JAR file so that it can be distributed easily.

jinnieshin committed:
Level 9

Implemented find command to find tasks that contains the string that the user wants to find.

Removed tabs from all messages.

jinnieshin committed:
Added JavaDoc

Added JavaDoc for major classes; Duke, Ui, Parser, TaskList, Task, Storage, FormatChecker, DukeException, Command.

jinnieshin committed:
Merge pull request #5 from jinnieshin/branch-A-JavaDoc

Added JavaDoc

jinnieshin committed:
Merge branch 'master' into branch-Level-9

jinnieshin committed:
Merge pull request #6 from jinnieshin/branch-Level-9

Level 9

jinnieshin committed:
Update FormatChecker.java

Updated FormatChecker to test formats based on different expected outputs

jinnieshin committed:
Merge pull request #7 from jinnieshin/branch-A-CodingStandard

Update FormatChecker.java

jinnieshin committed:
A-CheckStyle

Added checkstyle to detect coding style violation.

jinnieshin committed:
Fixed checkstyle errors

Fixed coding style violations excluding missing JavaDoc

gitsac commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,164 @@
+    /**
+     * Add a Deadline Task to the TaskList.
+     * @param input the content of the new Deadline Task
+     * @param deadline the deadline date and time of the new Deadline Task
+     * @param needPrint the indicator that shows whether the message needs to be printed
+     * @throws DukeException if the content is empty
+     */
+    public void add(String input, String deadline, boolean needPrint) throws DukeException {
+        if (isEmpty(input) || isEmpty(deadline)) {
+            throw new DukeException("The description and deadline of a deadline cannot be empty.");
+        }
+        Deadline newTask = new Deadline(input, deadline);
+        this.list.add(newTask);
+        if (needPrint) {
+            System.out.println("Got it. I've added this task:");
+            System.out.println(newTask.toString());
+            System.out.println("Now you have " + (this.list.size() - 1) + " tasks in the list.");
+        }
+    }
+
+    /**
+     * Add a Event Task to the TaskList.
+     * @param input the content of the new Event Task
+     * @param from the start date and time of the new Event Task
+     * @param to the end date and time of the new Event Task
+     * @param needPrint the indicator that shows whether the message needs to be printed
+     * @throws DukeException if the content is empty
+     */
+    public void add(String input, String from, String to, boolean needPrint) throws DukeException {
+        if (input.equals("") || from.equals("") || to.equals("")) {
+            throw new DukeException("The description and duration of an event cannot be empty.");
+        }
+        Event newTask = new Event(input, from, to);
+        this.list.add(newTask);
+        if (needPrint) {
+            System.out.println("Got it. I've added this task:");
+            System.out.println(newTask.toString());
+            System.out.println("Now you have " + (this.list.size() - 1) + " tasks in the list.");
+        }
+    }
```
Perhaps making the function names more descriptive in terms of which tasks are being added would be better?

gitsac reviewed (approved):
Overall, your iP has been very well-done! The commands are all well-abstracted, imports and JavaDocs are all done correctly, and the tests written for FormatChecker and Command are good!

vaidyanaath commented on src/main/java/duke/FormatChecker.java:
```diff
@@ -0,0 +1,43 @@
+        if (command.matches("^.+(\\s)/by(\\s).+$")) {
+            return true;
+        }
+        return false;
```
Perhaps you can simplify readability by using:
`return command.matches("^.+(\\s)/by(\\s).+$")`

jinnieshin replied:
Thank you for the suggestion! Just changed it :)

vaidyanaath commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,49 @@
+package duke;
+import duke.command.*;
```
Isn't it a better practice to import packages explicitly ?

vaidyanaath commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,49 @@
+        case "find":
+            return new FindCommand(content);
+        default:
+            return new UnknownCommand();
```
I like how you have used switch-case to keep the code concise and easy to read. Good job on following the correct code style indentation! 😄

vaidyanaath commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,91 @@
+            case "T":
+                try {
+                    tasks.add(taskData[2], false);
+                    if (taskData[1].equals("1")) {
```
In this line and below, the reader might not know what taskData[1], taskData[2], .. contains. Perhaps you can assign these values to descriptive variables before using them?

vaidyanaath commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,91 @@
+     * @throws IOException If file cannot be written into the data file.
+     */
+    public void writeToFile(String dukeData) throws IOException {
+        FileWriter fw = new FileWriter(this.filePath);
```
Very minor suggestion: the coding standard suggests that one doesn't use the `this` keyword unless necessary.

vaidyanaath commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,164 @@
+ * List of tasks that the user can add, mark, find or delete.
+ */
+public class TaskList {
+    private ArrayList<Task> list = new ArrayList<>();
```
Could the name for the `list` variable be more descriptive/ meaningful? The se-edu coding standard suggests using the plural form to represent collections.

vaidyanaath commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,164 @@
+     * @param needPrint the indicator that shows whether the message needs to be printed
+     * @throws DukeException if the content is empty
+     */
+    public void add(String input, boolean needPrint) throws DukeException {
```
Minor naming suggestion: Would it sound more like a boolean if you used `needsPrint` instead?

vaidyanaath commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,164 @@
+     * @param index the Task to be returns
+     * @return Task with the given index in the TaskList
+     */
+    public Task get(int index) {
```
The name for this function seems to be very similar to the `get` function in Arraylist and performs nearly the same operation. Would it be better to highlight the difference?

jinnieshin replied:
Thanks for noticing me! I changed its name to getTask.

vaidyanaath commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,164 @@
+     */
+    public TaskList() {
+        this.list.add(new ToDo("zeroth"));
+    }
```
Is there any reason why you decided to do this instead of manipulating the index values in relevant functions directly?

jinnieshin replied:
I felt more comfortable with not changing the indexes as I thought I may forget to manipulate the index sometimes.

vaidyanaath commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,41 @@
+        }
+        String[] dlTask = content.split("/by");
+        dlTask[0] = dlTask[0].trim();
+        dlTask[1] = dlTask[1].trim();
```
Would it be better if you renamed `dlTask` to be more readable?

vaidyanaath commented on src/main/java/duke/command/EventCommand.java:
```diff
@@ -0,0 +1,40 @@
+            tasks.add(eventTask[0], eventTask[1], eventTask[2], true);
+        } catch (DukeException e) {
+            throw e;
+        }
```
Do you need the try catch block here since you're throwing the exception anyways?

CedricPei replied:
try-catch block maybe no need cuz the exceptions are thrown

vaidyanaath commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,28 @@
+
+public class Event extends Task {
+    private LocalDateTime from;
+    private LocalDateTime to;
```
Perhaps you can change the name of the variables to show that they represent date/times?

vaidyanaath reviewed (commented):
I liked how you abstracted the different parts, especially the commands. I found the javadocs to be descriptive and well-formatted. The JUnit tests were comprehensive and covered a lot of cases. Overall, the iP was nicely done. Perhaps by changing a few variable names, it would be even better. Great work!

CedricPei commented on src/main/java/duke/FormatChecker.java:
```diff
@@ -0,0 +1,43 @@
+        }
+        return false;
+    }
+}
```
Wonderful idea for this FormatChecker class

CedricPei commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,49 @@
+            return new UnknownCommand();
+        }
+    }
+}
```
Great to extract the commands handling process to commands. It might be very helpful for future extension and development.

CedricPei commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,22 @@
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task {
+    private LocalDateTime deadline;
```
naming of this maybe follow camel

CedricPei commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,64 @@
+            System.out.println("Error during saving");
+        }
+    }
+}
```
Well done bro, instead of reviewing, I'm actually learning from your code.

jinnieshin committed:
Add GUI for Duke

Added GUI for Duke using JavaFX

Modified execute method of command classes to return String.

Added fxml files for styling

Need to add greeting and ending.

jinnieshin committed:
Merge pull request #8 from jinnieshin/branch-Level-10

Branch level 10

jinnieshin committed:
Update README.md

Added explanations on features of Duke.

jinnieshin committed:
Merge pull request #9 from jinnieshin/Use-GFMD

Updated README.md

jinnieshin committed:
Updated the code to meet the coding standard

Updated the code to meet the coding standard (except Javadoc, which would be updated soon)

jinnieshin committed:
Used Assertions to check correct format

Used Java Assertions to check whether formats for deadline command, event command and date is correct.

jinnieshin committed:
Merge pull request #11 from jinnieshin/branch-A-Assertions

Used Assertions to check correct format

jinnieshin committed:
Added Javadoc, Matched coding standard

Added Javadoc for classes and methods without Javadoc.

Removed unnecessary throwing statement of DukeException in execute method of some commands.

Refactored class name of TodoCommand to be ToDoCommand so that it follows the coding standard.

Changed the names of add method in TaskList so that it is easier to identify which task is being added to the TaskList.

Modified find method in TaskList to show the index to match the index of the task in the whole TaskList, not in the list of matching tasks.

jinnieshin committed:
Merge pull request #12 from jinnieshin/branch-A-CodeQuality

Added Javadoc, Matched coding standard

jinnieshin committed:
Added to detect duplicates, Fixed minor javafx issues

Added lines to detect duplicated tasks so that it would not be added multiple times.

Changed background colour of main window to match the dialog box.

Set maxHeight of dialog box to be max value so that it can show long content.

jinnieshin committed:
Changed name of the software, Fixed minor bugs

Renamed the software to DOggi, changed responses shown to the user to meet the concept.

Fixed the bug that does not show the full content when the content is long.

Added greeting message when the software starts.

Added execution of the software when EndCommand is executed.

Added more condition for commands involving index so that index smaller than 1 (out of range) throws an exception.

jinnieshin committed:
Updated user guide and product screenshot

Updated user guide and product screenshot for the github page.

jinnieshin committed:
Update Ui.png

jinnieshin committed:
Fixed minor bugs

jinnieshin committed:
Fixed file not found for jar file bug and minor bugs

Added file to be created if data file does not exists.

Renamed package GUI to gui so that it follows the coding standard.
